DESIGN.md

```diff
@@ -0,0 +1,58 @@
+# Module Design
+
+## Classes
+
+There are 4 main public classes and other can be added if they need to be managed separately than the existing classes:
+
+* sensu - Resources shared by all other public classes such as agent and backend
+* sensu::agent - Sensu Agent
+* sensu::backend - Sensu Backend
+* sensu::plugins - Sensu plugins
+
+Subclasses to the above hold resources for that resources to keep the logic in the main public class shorter. So `sensu::backend::default_resources` is a private classes that adds functionality to `sensu::backend`. There are also some shared classes like `sensu::ssl` that are private and have resources shared by other public classes.
+
+If a parameter is used by both agent and backend, it belongs in `sensu` class. All other parameters should be added to the appropriate public class.
+
+## Types/Provides
+
+### Sensuctl types
+
+The types to `sensuctl` resources are designed to closely match the resource specifications from Sensu Go. For example the Sensu Go `check` resource property of `interval` would map directly to the `sensu_check` type property of `interval`.
+
+Keys in the Sensu Go specification under `spec` should have a 1-to-1 mapping with Puppet type properties.
+
+The exception to the 1-to-1 mapping is Sensu Go resource `metadata`. The values for `metadata` keys of `name`, `namespace`, `labels`, and `annotations` are pulled out into Puppet type properties. The metadata exception is to properly handle resource names and namespaces and relationships as well as the composite names used by namespace capable resources.
+
```

Comment:
what resources are namespace capable and which are not? would a table be useful?

Reply:
I think this would cause too much duplication with what's provided by Puppet Strings and also Sensu Go docs. This would just become one more thing we have to maintain as Sensu Go evolves and when we update custom types. Should I still add something?